### DISCUSSION TITLE
missing python interface DumpFF2VTK

### DISCUSSION THREAD

**Discussion by LubomirJagos** [thliebig/openEMS-Project#151] [General]
Hi, in current python interface there is missing DumpFF2VTK function to be able generate 3D field vizualization from NF2FF function so I have reworte matlab/octave one and test it, here it is:
```python
import math

def DumpFF2VTK(farfield, t, a, filename):
    """
       Create .vtk file
       params:
           farfield: 2D array of values of field
           t:        theta angles in radians
           a:        phi angles in radians
           filename: output file name
    """

    with (open(filename, 'w') as outFile):
         outFile.write(f"# vtk DataFile Version 3.0\n")
         outFile.write(f"Structured Grid by python-interface of openEMS\n")
         outFile.write(f"ASCII\n")
         outFile.write(f"DATASET STRUCTURED_GRID\n")

         outFile.write(f"DIMENSIONS 1 {len(t)} {len(a)}\n")
         outFile.write(f"POINTS {len(t)*len(a)} double\n")

         for na in range(len(a)):
             for nt in range(len(t)):
                 val1 = farfield[nt][na]*math.sin(t[nt])*math.cos(a[na])
                 val2 = farfield[nt][na]*math.sin(t[nt])*math.sin(a[na])
                 val3 = farfield[nt][na]*math.cos(t[nt])
                 outFile.write(f"{val1} {val2} {val3}\n")

         outFile.write(f'\n\n');
         outFile.write(f'POINT_DATA {len(t)*len(a)}\n');
         outFile.write(f'SCALARS gain double 1\n');
         outFile.write(f'LOOKUP_TABLE default\n');
         for na in range(len(a)):
             for nt in range(len(t)):
                 outFile.write(f"{farfield[nt][na]}\n")
```

**Comment by oberstet** [comment]
this is great, love it!

I want to try that using a most simple model - a **center-fed half wave dipole** - which is well understood and for which I can already compute a number of interesting data/plots with openEMS, and I could (basically) verify values expected from theory - which is what creates "trust" for me.

I have now added some some directivity measures, however *I do not fully understand those pieces yet though*.

Hence, I don't have a valid/trusted basis yet to add _even more_, like a 3D radiation plot! Which I do want;)

In any case, I will do some experimentation, and I also post complete code and results below.

----

**Code**: complete working and self-contained code is here https://gist.github.com/oberstet/ce57c9579591d6d1b1f9222afa00bccc

**Results**:

![Figure_2](https://github.com/thliebig/openEMS-Project/assets/233340/6e795bfd-f5d6-478d-ae7d-50fd3c04a7fe)

![Figure_1](https://github.com/thliebig/openEMS-Project/assets/233340/211cf55f-5ac7-42e2-adb1-4b88988e1097)

![Figure_3](https://github.com/thliebig/openEMS-Project/assets/233340/4cf6ce14-7435-4517-90aa-da368cca30af)

---

Open questions (well, "open" for me;):

1. One piece in particular that I do not understand is the selection of frequencies (pls see https://gist.github.com/oberstet/ce57c9579591d6d1b1f9222afa00bccc#file-dipole-py-L417 .. also, the results in the comments there ... how can "Efficiency: nu_rad" be >100%?), the mesh resolution and the radius for NF2FF recording box / computation
2. Another piece that I have trouble with is "theta" and "phi" .. my dipole is oriented along the X-axis. I don't know if that is "common", and in any case, I haven't found a definition of theta/phi in the [docs](https://docs.openems.de/python/openEMS/nf2ff.html#openEMS.nf2ff.nf2ff.CalcNF2FF):

```
# CalcNF2FF(sim_path, freq, theta, phi, radius=1, center=[0, 0, 0], outfile=None, read_cached=False, verbose=0)
# theta/phi – array like – Theta/Phi angles to calculate the far-field
```

Well yes, and that's what it is used _for_, but what is the _definition_ of Theta/Phi? Obviously, it could mean "theta = angle between FF and Z axis" .. or it could mean sth else.

Ultimately, I want to render a picture like the following (which aligns the dipole along Z axis, not X like I do):

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/50366675-3aff-41b9-b4be-c0cffe1dc394)

**Reply by VolkerMuehlhaus** [reply]
If radiation efficiency is slightly >100%, this might result from a lumped port that is too large, so that the voltage probe inside the port is affected by the external fields. Try to use a shorter lumped port and see if that solves the issue.

Regarding theta and phi, there is no general rule what orientation for your dipole should be used. For directed antennas, we usually want to have the main beam in z direction, at theta=0°.

**Reply by oberstet** [reply]
Thank you so much for your explanations! Worth gold;) 

> lumped port that is too large

ok, I will try! currently, I use a small (0.5mm) overlap, but not zero, as zero doesn't work. I figured the lumped port only touching the dipole arms isn't enough. anyways, I will reduce it and check.

>  For directed antennas, we usually want to have the main beam in z direction, at theta=0°.

np, I can rearrange the geometry to have the dipole along Z.

"theta=0": ok, above diagram b) has a *visual definition* of theta with the dipole along Z (which you suggest), if that matches a definition of theta you have in mind, then I can work from that! thanks=) [yes, I did RTFM /  the docs, but didn't find any definitions of theta or phi ..)

**Reply by VolkerMuehlhaus** [reply]
With lumped port size I meant the size of the port itself. In the picture that you showed (CST?) the port is more than 10% of the total dipole length. In my experience, a much shorter lumped port works best.

Regarding the overlap between ports and metals: I always use zero overlap. If the metal overlaps parts of the port, this can give incorrect results, because the measured voltages and the termination resistor are partially shorted.

**Reply by oberstet** [reply]
In the screenshot above which shows the geometry I have indeed used a larger lumped port, just to make it clearly visible (the arrangement of the geometry). In the plots with S11 above, and for the value with efficiency >100% however, I used these parameters:

1. `unit = 1e-3`, `f0 = 446.1e6` and `fc = 0.15 * f0`
2. `dipole_wire_radius = 0.001`
3. `dipole_length = lambda0 / 2` with `lambda0 = opt_factor * C0 / f0 / unit` and `opt_factor = 0.8651`
4. `dipole_gap = 0.0`
5. `feed_resistance = 71.0`
6. `feed_overlap = 0.5`
7. `feed_radius = dipole_wire_radius / math.sqrt(2)`

pls see line https://gist.github.com/oberstet/3ec68a8348494452c2347764f1920bdd#file-dipole3-py-L75 and ff

so actually, 0.5 (mm) is much less than 1% of the resulting `Dipole (lambda/2) length is 290.7 mm`

----

I had a lot of problem with the whole lumped port setup stuff .. orientation, amount of overlap, exciting 1d wires, 2d shapes, or 3d wire etc etc

This stuff is poorly documented in my eyes, but yes, I am absolutely new to this stuff;) so kinda expected.

Anyways, this is kinda annoying me, and I want to have it out of my way, so tomorrow, I am going to add a database to store results and parameter sets, and then I can just "brute force" compute S11 and efficiency for resonance f for a whole bunch of parameters ... while sleeping;) And then I know how it behaves.

But your explanations are highly appreciated in any case (even if I have the systematic parameter search results), since it provides me _clues_ for actual _explanations_ ! So thank you a lot!

**Reply by VolkerMuehlhaus** [reply]
Hi Tobias,

I don't think that trial & error will help much to understand the settings.

Your model is somewhat confusing to me - you have defined 1µm wire diameter, is that what you wanted? This makes it really difficult to see anything useful in AppCSXCAD, so I increased wire diameter to 2mm for debugging purposes. For the dipole_gap I used 1mm and set your feed_overlap to zero.

With these settings, we can see in **2D side view** that your mesh is very coarse in the port region, and doesn't align with the ends of the metal or port. This is why it didn't work without cheating by the oversize value.

![port_sideview](https://github.com/thliebig/openEMS-Project/assets/47922981/805f8d91-5676-4032-8358-cb29f6d61d9e)

I would use a different mesh strategy here: something like 4 mesh lines across the length of the port, and also dense mesh near the end of the dipole arms. In between, you can use a much coarser mesh up to lambda/20.

If you can explain your modelling task (resolve some finite dipole diameter or just simulate a thin wire?), I can then try to help with meshing.

Best regards
Volker

**Reply by VolkerMuehlhaus** [reply]
To explain the graded mesh strategy that I would use, I have attached an example. This is essentially your model, with the changes that I mentioned above, and different meshing in x direction. 

One other minor change: in my opinion the port gap is part of the total dipole length, so I changed the dipole arm length code.

Model is attached. Note that my code ends after showing the model in AppCSXCAD, I removed the calculation part because I only wanted to visualize the model and mesh.

![gradedmesh](https://github.com/thliebig/openEMS-Project/assets/47922981/a0faacec-7a66-4b99-b024-2c0a0ea1ea2e)

[dipole3_reduced_volker.zip](https://github.com/thliebig/openEMS-Project/files/13480239/dipole3_reduced_volker.zip)

**Reply by oberstet** [reply]
This is great feedback, and your explanations help me a lot!! And the fact that you even provided demo code is fantastic=) Thank you so much! I will need to digest that, and try your code.

---

> your mesh is very coarse in the port region, and doesn't align with the ends of the metal or port. This is why it didn't work without cheating by the oversize value.

aaahhh, right! got it.

---

I am really missing some basics. Which isn't a surprise, as I am completely new to RF. I am coming from a different world: digital / software. I've got quite a lot of know how and experience there, also in open source, and also in Python specifically, which is one reason I was naturally drawn to openEMS, with the Python interface.

The way and reasons of why I ended here (creating some handheld radio etc) are strange-crazy, but too long to lay out here, and they don't matter (if you want to know, drop me a PM, I can share).

Now, being a developer, I am also kinda "lazy" ... doing things manually isn't my thing ... why not write a program that writes a program or searches through parameters? Why should I fiddle around manually for hours? But you are of course right with your hint that I might get results, but no explanation or insight. True.

I learned that already the hard way .. e.g. I couldn't get any dipole working initially .. there is no example .. and I was trying to figure out how to position, connect and drive _two_ lumped ports - because a dipole has two antenna elements, right? What I was missing is: even a monopole also has two elements, one of which just turns out to be something acting as a ground plane, a reflector, and the goal is to make the reflected antenna image constructively overlay the primary image. Well, ok, then of course 1 lumped port in both cases makes a lot more sense;) Basics, I know.

> Your model is somewhat confusing to me - you have defined 1µm wire diameter, is that what you wanted? 

No, this whole dipole thing is just to establish the most simple antenna as a reference, and fully understand that, before moving to my actual goal of auto-optimizing an antenna (during which the reference will come in handy).

> If you can explain your modelling task (resolve some finite dipole diameter or just simulate a thin wire?), I can then try to help with meshing.

my actual goal is

*A PCB antenna (4-layer FR4) for the whole range ~400-950 MHz, omnidirectional, for a handheld radio, and with antenna size  < ~ 70x10mm (thus, ~ lambda/8).*

I initially thought: can't be too hard, should be done in 2 weeks;) Yes, hyperbole, I now realize it isn't that simple .. for various reasons. But I'm making progress, and if so only incrementally, I will reach the goal, I'm sure;)

So here a pic drawn in KiCad where I started from 2 weeks ago, and this layout is just me taking a wild guess;) By now, I'm not sure a planar helical will cut it, mostly due to bandwidth, anyways, once I got the auto-optimizer controlling a suitable geometry generator, I should be able to get more sleep;)

![xyz](https://github.com/thliebig/openEMS-Project/assets/233340/da1befec-99bc-4c70-8424-5e8237f9a02e)

**Reply by oberstet** [reply]
alright, I was interrupted by sth else, but now I got back:

* tested your code verbatim: works!
* I've analyzed and integrated all your changes: works!
* I've re-optimized the opt/correction factor (because the changed dipole wire radius and gap) - the new value is a tiny bit below the old value (and also the impedance is 0.2 ohm lower):
* new code is here: https://gist.github.com/oberstet/8877417387c622f59700612818a1484d

 *take away: failing on proper mesh setup can evoke strange artifacts, like with my port overlap stuff. I learned something;) thanks a lot!*

```
# use with dipole_wire_radius = 0.001
# opt_factor = 0.8651

# use with dipole_wire_radius = 2.0
# Found resonance frequency at 446.1 MHz with -91.6 dB at 70.8 Ohm
# Dipole (lambda/2) length is 285.8 mm
opt_factor = 0.8506
```

the optimizer aims to hit a **fixed point** of the S11 function, that is, it hits an `opt_factor` _and_ feed `impedance` such that the simulated/computed resonance frequency and impedance _for the excitation frequency_ (which is set to the center of the desired band) is identical to the former values.

*I have no clue if this is "usually" done, or what, and what it means, but for me, it was natural (from a pure math angle) to aim for a fixed point, and practically, it achieves quite good results (see below)*


```
================================================================================

Found resonance frequency at 446.1 MHz with -91.6 dB at 70.8 Ohm
Dipole (lambda/2) length is 285.8 mm

================================================================================

S11 at frequency 446.0 MHz is -64.4 dB at 70.8 Ohm [index 999]
S11 at frequency 446.1 MHz is -91.6 dB at 70.8 Ohm [index 1000]
S11 at frequency 446.2 MHz is -65.1 dB at 70.8 Ohm [index 1001]


S11 at frequency 412.6 MHz is -10.0 dB at 54.3 Ohm
S11 at frequency 492.9 MHz is -10.0 dB at 102.1 Ohm

S11 at frequency 426.9 MHz is -15.0 dB at 60.7 Ohm
S11 at frequency 469.0 MHz is -15.0 dB at 85.0 Ohm

S11 at frequency 435.0 MHz is -19.9 dB at 64.8 Ohm
S11 at frequency 458.3 MHz is -20.0 dB at 78.0 Ohm

S11 at frequency 439.7 MHz is -24.9 dB at 67.3 Ohm
S11 at frequency 452.8 MHz is -25.0 dB at 74.7 Ohm

S11 at frequency 442.5 MHz is -29.9 dB at 68.8 Ohm
S11 at frequency 449.8 MHz is -29.9 dB at 73.0 Ohm

S11 at frequency 444.9 MHz is -39.5 dB at 70.1 Ohm
S11 at frequency 447.3 MHz is -39.7 dB at 71.5 Ohm

S11 at frequency 445.7 MHz is -49.1 dB at 70.6 Ohm
S11 at frequency 446.5 MHz is -49.2 dB at 71.0 Ohm
================================================================================
```

![Figure_2](https://github.com/thliebig/openEMS-Project/assets/233340/2ff2c3f3-3010-4096-9ebe-7a1c4adb89d0)

----

finally I also oriented the dipole along Z-axis, and I am now figuring out this directivity stuff and hopefully will also be able to render a 3D radiation pattern soon

what does this plot tell me?

![Figure_3](https://github.com/thliebig/openEMS-Project/assets/233340/6d8c9de4-63c4-452b-ab8b-d292d7749e0f)

**Reply by VolkerMuehlhaus** [reply]
Hi Tobias,

maybe you should open a new thread, as we are off-topic and there are many items to discuss.

For me, your code approach looks somewhat unusual. When I design PCB antennas, I use wideband excitation for FDTD simulation, that works fine and is much easier than your approach. 

Extracting the resonance frequency by code is fine, but overall your model code is somewhat overloaded for the basic steps that we need to cover now. I was unable to get your latest code up and running here, so not sure what goes wrong.

One important detail is mesh density around the radiator/wire, in your last model that was much too coarse except for meshing in longitudinal wire direction. The fields at the end of the wire was modelled with big rectangles (large mesh size perpendicular to the wire) that are much too large to resolve the actual field distribution at the open end (field gradients are very strong there!)

Regarding the pattern that you show above: that is a 2D cut of the antenna pattern, not sure what direction (what cutting plane) we see here. Dipole directivity should peak at ~ 2.15 dBi in the maximum of the "donut" antenna pattern, so the -40dBi look unexpected.

But let's open another thread, and maybe someone else can jump in to support you.

Best regards
Volker

**Reply by oberstet** [reply]
+1! sorry for OT noise. https://github.com/thliebig/openEMS-Project/discussions/152

**Comment by oberstet** [comment]
alright, I've added above function and the VTK file dump at the end to above example. full updated source is here:

https://gist.github.com/oberstet/8e771e2466395d6f164d93fe8f7a7d32

this runs into the following error:

```
Traceback (most recent call last):
  File "/home/oberstet/scm/typedefint/rfminer/_work/poc_test_dipole2.py", line 577, in <module>
    directivity = nf2ff.Prad[0] / nf2ff.Prad * 4 * math.pi
                  ^^^^^^^^^^
AttributeError: 'nf2ff' object has no attribute 'Prad'
```

as the nf2ff object attribute dump https://gist.github.com/oberstet/8e771e2466395d6f164d93fe8f7a7d32#file-dipole2-py-L571 above shows, the nf2ff object neither has a `Prad` nor a `P_rad` attribute, but only these attributes:

```
['CSX',
 'CalcNF2FF',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'center',
 'directions',
 'dump_mode',
 'dump_type',
 'e_dump',
 'e_file',
 'freq',
 'h_dump',
 'h_file',
 'mirror',
 'name',
 'phi',
 'start',
 'stop',
 'theta']
```

**Comment by oberstet** [comment]
I just recognize that you are using `nf2ffBoxList` to create the NF2FF recording box, as in

```python
#
#	nf2ffBoxList[<name>] - NF2FF box structure which should be calculated
#       INPUT ANGLES ARE IN DEGREES for python interface!
#
nf2ff = nf2ffBoxList['nf2ffbox_1_-_farfield_1'].CalcNF2FF(Sim_Path, plotFrequency, thetaRange, phiRange, outfile='3D_Pattern.h5', verbose=True, read_cached=False)
```

whereas I am using

```python
nf2ff = fdtd.CreateNF2FFBox("nf2ff-box", start=start, stop=stop, opt_resolution=[mesh_res_farfield] * 3)
```

**Where does nf2ffBoxList come from?**

**Comment by oberstet** [comment]
alright, is seems I've got that fixed https://gist.github.com/oberstet/3ec68a8348494452c2347764f1920bdd

```
...
================================================================================


Calculating the 3D far field...
Analyzing far-field at radius 126
Analyzing far-field for 1205 frequencies:
array([4.12843245e+08, 4.12910160e+08, 4.12977075e+08, ...,
       4.93275075e+08, 4.93341990e+08, 4.93408905e+08])
nf2ff: Reading planes: /tmp/dipole/nf2ff-box_E_0.h5 & /tmp/dipole/nf2ff-box_H_0.h5
nf2ff: Data-Size: 1x7x7
nf2ff: Analysing far-field for 1205 frequencies.  
nf2ff: Reading planes: /tmp/dipole/nf2ff-box_E_1.h5 & /tmp/dipole/nf2ff-box_H_1.h5
nf2ff: Data-Size: 1x7x7
nf2ff: Analysing far-field for 1205 frequencies.  
nf2ff: Reading planes: /tmp/dipole/nf2ff-box_E_2.h5 & /tmp/dipole/nf2ff-box_H_2.h5
nf2ff: Data-Size: 8x1x7
nf2ff: Analysing far-field for 1205 frequencies.  
nf2ff: Reading planes: /tmp/dipole/nf2ff-box_E_3.h5 & /tmp/dipole/nf2ff-box_H_3.h5
nf2ff: Data-Size: 8x1x7
nf2ff: Analysing far-field for 1205 frequencies.  
nf2ff: Reading planes: /tmp/dipole/nf2ff-box_E_4.h5 & /tmp/dipole/nf2ff-box_H_4.h5
nf2ff: Data-Size: 8x7x1
nf2ff: Analysing far-field for 1205 frequencies.  
nf2ff: Reading planes: /tmp/dipole/nf2ff-box_E_5.h5 & /tmp/dipole/nf2ff-box_H_5.h5
nf2ff: Data-Size: 8x7x1
nf2ff: Analysing far-field for 1205 frequencies.  
Radiated power: P_rad = 3.3366863684313314e-21 W
Directivity: D_max = -42.45806364612437 dBi
Efficiency: nu_rad = 24.669637065384926 %
Theta_HPBW = 43.0 °
['CSX',
 'CalcNF2FF',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'center',
 'directions',
 'dump_mode',
 'dump_type',
 'e_dump',
 'e_file',
 'freq',
 'h_dump',
 'h_file',
 'mirror',
 'name',
 'phi',
 'start',
 'stop',
 'theta']
(cpy311_7) oberstet@intel-nuci7:~/scm/typedefint/rfminer$ tree /tmp/dipole/
/tmp/dipole/
├── 3D_Pattern_CPLH.vtk
├── 3D_Pattern_CPRH.vtk
├── 3D_Pattern_E_norm.vtk
├── 3D_Pattern.vtk
├── et
├── ht
├── nf2ff-box_E_0.h5
├── nf2ff-box_E_1.h5
├── nf2ff-box_E_2.h5
├── nf2ff-box_E_3.h5
├── nf2ff-box_E_4.h5
├── nf2ff-box_E_5.h5
├── nf2ff-box_H_0.h5
├── nf2ff-box_H_1.h5
├── nf2ff-box_H_2.h5
├── nf2ff-box_H_3.h5
├── nf2ff-box_H_4.h5
├── nf2ff-box.h5
├── nf2ff-box_H_5.h5
├── port_it_1
└── port_ut_1

0 directories, 21 files
(cpy311_7) oberstet@intel-nuci7:~/scm/typedefint/rfminer$ 
```

----

let me now check paraview ..  I guess ..

**Reply by LubomirJagos** [reply]
3D_pattern.vtk is radiation pattern
3D_Pattern_E_norm.vtk is electric field

**Reply by oberstet** [reply]
this is what I've got (the output generated by above script)  

```
(cpy311_7) oberstet@intel-nuci7:~/scm/typedefint/rfminer$ du -hs /tmp/dipole/3D_Pattern.vtk
2,4M	/tmp/dipole/3D_Pattern.vtk
(cpy311_7) oberstet@intel-nuci7:~/scm/typedefint/rfminer$ du -hs /tmp/dipole/*.vtk
2,4M	/tmp/dipole/3D_Pattern_CPLH.vtk
2,4M	/tmp/dipole/3D_Pattern_CPRH.vtk
2,9M	/tmp/dipole/3D_Pattern_E_norm.vtk
2,4M	/tmp/dipole/3D_Pattern.vtk
```

how do I make paraview digest that and show something that makes sense?

----

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/3e869171-bcac-4ff9-9d9a-84fc9e1a8cde)

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/08d1510e-236d-493c-8821-2c0ece3110fa)

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/fa7600c0-d129-43b4-a9ec-4dbc234eef3b)

**Comment by oberstet** [comment]
mmmh .. I mean, it *does* look interesting;) *what do I need to do in paraview, which of the 4 files to open? what rendering/filter settings to set?* sorry, I'm pretty much a paraview newbie ...

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/c5ccc9a9-14a2-4742-ac6f-289689727294)

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/50e7698a-1c83-4445-b855-b8d414fca163)
